### PR TITLE
Refactor DATE, TIME, TIMESTAMP/TZ in strconv

### LINKED ISSUE #2348: Support iso-formatted timestamp strings
For example, a query that works in PostgreSQL is:
```
jessicalaughlin=# SELECT EXTRACT(EPOCH FROM TIMESTAMP '2020-03-16T22:34:42.846078');
     date_part
-------------------
 1584398082.846078
(1 row)
```

But in Materialize, we see:
```
materialize=> SELECT EXTRACT(EPOCH FROM TIMESTAMP '2020-03-16T22:34:42.846078');
ERROR:  Invalid TIMESTAMP '2020-03-16T22:34:42.846078': invalid DateTimeField: T
```

### PR DESCRIPTION
**Pull request by sploiselle**
Includes:

- Support for a subset of [ISO 8601](https://en.wikipedia.org/wiki/ISO_8601) timestamps, which primarily includes allowing users to separate dates and times with the character `T`. (We are not yet supporting timestamps like `20200318T153032Z`.) (#2348)
- Documentation to reflect this change, as well the truth of our more permissive parsing of `date` and `timestamp` data.
- Support an arbitrary number of spaces between dates, times, intervals, and `DateTimeDelimeter`, e.g. `TIMESTAMP '2001-01-01 T 01:02'` but with a lot of spaces.
- Refactor of `src/repr/scalar/datetime.rs` to use `VecDeque` instead of passing around slices and occasionally making them `peekable`; it became very unergonomic when wanting to share data between methods that fill `ParsedDateTime`s (`fill_pdt...`).
- Push the complexity for parsing and generating `ParsedDateTime` into `src/repr/scalar/datetime.rs`; there was more in `src/repr/scalar/strconv.rs` than felt right.
- Actually parse (and error on) timezone strings, irrespective of type (#2358)

### Questions

- How can I structure function calls on a struct that don't perform a move when returning a new value? [Example](https://github.com/MaterializeInc/materialize/compare/master...sploiselle:iso-timestamp?expand=1#diff-10ded2b16b6de021f4b12e1cd75a0640R194-R195)
- I cannot figure out where `fuzz/artifacts/fuzz_sqllogictest/.gitignore` came from, and I am failing a seemingly related test; any idea what's going on here?

Fix #2348 #2358

### CODE REVIEW

File: src/repr/scalar/datetime.rs [hunk 1]
```diff
@@ -6,11 +6,14 @@
 // As of the Change Date specified in that file, in accordance with
 // the Business Source License, use of this software will be governed
 // by the Apache License, Version 2.0.
-
+use std::collections::VecDeque;
```
Comment (benesch) [inline src/repr/scalar/datetime.rs]: Minor, but newline above the first `use`

File: src/repr/scalar/datetime.rs [hunk 2]
```diff
@@ -6,11 +6,14 @@
...
+use std::convert::TryInto;
 use std::fmt;
 use std::str::FromStr;
 use std::time::Duration;
 
+use chrono::{NaiveDate, NaiveTime};
+
```
Comment (benesch) [inline src/repr/scalar/datetime.rs]: And this should be in the same block as `failure`.

File: src/repr/scalar/datetime.rs [hunk 3]
```diff
@@ -1006,33 +1118,31 @@ fn expected_sql_standard_interval_tokens(from: DateTimeField) -> Vec<TimeStrToke
...
+    let (start, end) = match from {
+        Year => (0, 4),
+        Month => (2, 4),
+        Day => (4, 6),
+        hms => {
+            return expected_dur_like_tokens(hms);
+        }
+    };
+
+    Ok(VecDeque::from(all_toks[start..end].to_vec()))
```
Comment (benesch) [inline src/repr/scalar/datetime.rs]: And then drop the `Ok`  here.

File: src/repr/scalar/datetime.rs [hunk 4]
```diff
@@ -429,36 +482,18 @@ impl ParsedDateTime {
     ///
```
[the remark this answers is not in the document]
Comment (benesch) [reply src/repr/scalar/datetime.rs]: (I.e., the comment that is just out of frame.)

**Comment by benesch** [comment]
Btw, you have one actual SLT failure:

```

==> test/sqllogictest/interval.slt
--
  | SELECT INTERVAL '1-2 hour';
  | PlanFailure: Invalid INTERVAL '1-2 hour': Invalid syntax: HOUR must be preceeded by a number, e.g. '1HOUR' (ErrorMessage { msg: "Invalid INTERVAL \'1-2 hour\': Invalid syntax: HOUR must be preceeded by a number, e.g. \'1HOUR\'" })
```

which appears to just be an error message that changed.

**Comment by sploiselle** [comment]
@benesch tysm; will avail have tidier commits.

Also, is there a way to avoid clone [here](https://github.com/MaterializeInc/materialize/blob/075f1d0cc6f48dabe64cc37c6a229e1461304b89/src/repr/scalar/strconv.rs#L194)? Using [`TryInto for NaiveDate`](https://github.com/MaterializeInc/materialize/blob/075f1d0cc6f48dabe64cc37c6a229e1461304b89/src/repr/scalar/datetime.rs#L193) and [`TryInto for NaiveTime`](https://github.com/MaterializeInc/materialize/blob/075f1d0cc6f48dabe64cc37c6a229e1461304b89/src/repr/scalar/datetime.rs#L213). It feels like the move to `pdt` is avoidable but my efforts were bested.

**Comment by benesch** [comment]
You can implement `TryInto` for `&ParsedDateTime` instead of `ParsedDateTime` instead. But popping up a level: I don't think this is a good use of `TryInto`. In fact I'm very much suspicious of every implementation of `TryFrom`/`TryInto`. It's just a really verbose way of writing a method that takes a `ParsedDateTime` and produces a `NaiveDateTime`.

Proposal incoming momentarily.

**Comment by benesch** [comment]
```diff
diff --git a/src/repr/scalar/datetime.rs b/src/repr/scalar/datetime.rs
index fa3b1776..e6ab8682 100644
--- a/src/repr/scalar/datetime.rs
+++ b/src/repr/scalar/datetime.rs
@@ -190,16 +190,11 @@ impl Default for ParsedDateTime {
     }
 }
 
-impl TryInto<chrono::NaiveDate> for ParsedDateTime {
-    type Error = failure::Error;
-    fn try_into(self) -> Result<chrono::NaiveDate> {
+impl ParsedDateTime {
+    pub fn compute_date(&self) -> Result<chrono::NaiveDate> {
         match (self.year, self.month, self.day) {
             (Some(year), Some(month), Some(day)) => {
-                let p_err = {
-                    |e: std::num::TryFromIntError, field: &str| {
-                        failure::format_err!("{} in date is invalid: {}", field, e)
-                    }
-                };
+                let p_err = |e, field| format_err!("{} in date is invalid: {}", field, e);
                 let year = year.unit.try_into().map_err(|e| p_err(e, "Year"))?;
                 let month = month.unit.try_into().map_err(|e| p_err(e, "Month"))?;
                 let day = day.unit.try_into().map_err(|e| p_err(e, "Day"))?;
@@ -208,17 +203,9 @@ impl TryInto<chrono::NaiveDate> for ParsedDateTime {
             (_, _, _) => bail!("YEAR, MONTH, DAY are all required"),
         }
     }
-}
 
-impl TryInto<chrono::NaiveTime> for ParsedDateTime {
-    type Error = failure::Error;
-
-    fn try_into(self) -> Result<chrono::NaiveTime> {
-        let p_err = {
-            |e: std::num::TryFromIntError, field: &str| -> failure::Error {
-                failure::format_err!("invalid {} : {}", field, e)
-            }
-        };
+    pub fn compute_time(&self) -> Result<chrono::NaiveTime> {
+        let p_err = |e, field| format_err!("invalid {}: {}", field, e);
         let hour = match self.hour {
             Some(hour) => hour.unit.try_into().map_err(|e| p_err(e, "HOUR"))?,
             None => 0,
@@ -240,9 +227,7 @@ impl TryInto<chrono::NaiveTime> for ParsedDateTime {
         };
         Ok(NaiveTime::from_hms_nano(hour, minute, second, nano))
     }
-}
 
-impl ParsedDateTime {
     /// Compute an Interval from an IntervalValue. This could be adapted
     /// to `impl TryFrom<IntervalValue> for Interval` but is slightly more
     /// ergononmic because it doesn't require exposing all of the private
diff --git a/src/repr/scalar/strconv.rs b/src/repr/scalar/strconv.rs
index 56018817..b71b430b 100644
--- a/src/repr/scalar/strconv.rs
+++ b/src/repr/scalar/strconv.rs
@@ -23,7 +23,6 @@
 //! string representations for the corresponding PostgreSQL type. Deviations
 //! should be considered a bug.
 
-use std::convert::TryInto;
 use std::{f32, f64};
 
 use chrono::offset::TimeZone;
@@ -191,8 +190,8 @@ fn parse_timestamp_string(s: &str) -> Result<(NaiveDate, NaiveTime, i64), failur
     let (ts_string, tz_string) = crate::datetime::split_timestamp_string(s);
 
     let pdt = ParsedDateTime::build_parsed_datetime_timestamp(&ts_string)?;
-    let d: NaiveDate = pdt.clone().try_into()?;
-    let t: NaiveTime = pdt.try_into()?;
+    let d: NaiveDate = pdt.compute_date()?;
+    let t: NaiveTime = pdt.compute_time()?;
 
     let offset = if tz_string.is_empty() {
         0
@@ -228,7 +227,7 @@ where
 /// ```
 pub fn parse_time(s: &str) -> Result<NaiveTime, failure::Error> {
     match ParsedDateTime::build_parsed_datetime_time(&s) {
-        Ok(pdt) => pdt.try_into(),
+        Ok(pdt) => pdt.compute_time(),
         Err(e) => bail!("Invalid TIME '{}': {}", s, e),
     }
 }
```

**Pull request merged**